ocaml/libs/cpuid/cpuid.ml

```diff
@@ -0,0 +1,81 @@
+(*
+ * Copyright (C) Cloud Software Group.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+let string_of_features features =
+  Array.map (Printf.sprintf "%08Lx") features
+  |> Array.to_list
+  |> String.concat "-"
+
+exception InvalidFeatureString of string
+
+let features_of_string str =
+  let scanf fmt s = Scanf.sscanf s fmt (fun x -> x) in
+  try
+    String.split_on_char '-' str
+    |> (fun lst -> if lst = [""] then [] else lst)
+    |> Array.of_list
+    |> Array.map (scanf "%08Lx%!")
+  with _ -> raise (InvalidFeatureString str)
+
+(** If arr0 is shorter than arr1, extend arr0 with elements from arr1 up to the
+ *  length of arr1. Otherwise, truncate arr0 to the length of arr1. *)
+let extend arr0 arr1 =
```

Comment:
Is the name misleading? The code returns a new array that is some kind of alignment of `arr0` with `arr1`; this is not necessarily an extension but could be a truncation as well. Could be called `align arr0 ~with`; `pad` could be another name but it implies adding elements.

The semantics of this operation seems strange in the case where data from arr1 is taken to pad arr0. Maybe it would be better to call this step normalisation. All feature sets are normalised before doing arithmetic on them,

Reply:
The code in cpuid.ml is just moved from cpuid_helpers.ml to a library. I don't really want to change it if not necessary, because it is known to work, and will eventually go away.